[2024-07-01T12:28:17.486+0000] {local_task_job_runner.py:120} INFO - ::group::Pre task execution logs
[2024-07-01T12:28:17.547+0000] {taskinstance.py:2076} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: purchase_dag.kafka_producer manual__2024-07-01T12:28:14.869355+00:00 [queued]>
[2024-07-01T12:28:17.565+0000] {taskinstance.py:2076} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: purchase_dag.kafka_producer manual__2024-07-01T12:28:14.869355+00:00 [queued]>
[2024-07-01T12:28:17.565+0000] {taskinstance.py:2306} INFO - Starting attempt 1 of 1
[2024-07-01T12:28:17.586+0000] {taskinstance.py:2330} INFO - Executing <Task(PythonOperator): kafka_producer> on 2024-07-01 12:28:14.869355+00:00
[2024-07-01T12:28:17.597+0000] {logging_mixin.py:188} WARNING - /home/***/.local/lib/python3.12/site-packages/***/task/task_runner/standard_task_runner.py:61 DeprecationWarning: This process (pid=592) is multi-threaded, use of fork() may lead to deadlocks in the child.
[2024-07-01T12:28:17.598+0000] {standard_task_runner.py:63} INFO - Started process 593 to run task
[2024-07-01T12:28:17.596+0000] {standard_task_runner.py:90} INFO - Running: ['***', 'tasks', 'run', 'purchase_dag', 'kafka_producer', 'manual__2024-07-01T12:28:14.869355+00:00', '--job-id', '488', '--raw', '--subdir', 'DAGS_FOLDER/purchase_dag.py', '--cfg-path', '/tmp/tmp8i89haiy']
[2024-07-01T12:28:17.599+0000] {standard_task_runner.py:91} INFO - Job 488: Subtask kafka_producer
[2024-07-01T12:28:17.619+0000] {logging_mixin.py:188} WARNING - /home/***/.local/lib/python3.12/site-packages/***/settings.py:195 DeprecationWarning: The sql_alchemy_conn option in [core] has been moved to the sql_alchemy_conn option in [database] - the old setting has been used, but please update your config.
[2024-07-01T12:28:17.675+0000] {task_command.py:426} INFO - Running <TaskInstance: purchase_dag.kafka_producer manual__2024-07-01T12:28:14.869355+00:00 [running]> on host 6d5f7889ee5a
[2024-07-01T12:28:17.803+0000] {taskinstance.py:2648} INFO - Exporting env vars: AIRFLOW_CTX_DAG_OWNER='***' AIRFLOW_CTX_DAG_ID='purchase_dag' AIRFLOW_CTX_TASK_ID='kafka_producer' AIRFLOW_CTX_EXECUTION_DATE='2024-07-01T12:28:14.869355+00:00' AIRFLOW_CTX_TRY_NUMBER='1' AIRFLOW_CTX_DAG_RUN_ID='manual__2024-07-01T12:28:14.869355+00:00'
[2024-07-01T12:28:17.805+0000] {taskinstance.py:430} INFO - ::endgroup::
[2024-07-01T12:28:17.833+0000] {logging_mixin.py:188} INFO - {'item_name': 'Laptop', 'date_of_purchase': '2023-01-01', 'cost': 1200.0, 'quantity': 1, 'purchased_by': 'Father', 'payment_method': 'Card'}
[2024-07-01T12:28:17.835+0000] {logging_mixin.py:188} INFO - {'item_name': 'Laptop', 'date_of_purchase': '2023-01-01', 'cost': 1200.0, 'quantity': 1, 'purchased_by': 'Father', 'payment_method': 'Card'}
[2024-07-01T12:28:17.836+0000] {logging_mixin.py:188} INFO - {'item_name': 'Smartphone', 'date_of_purchase': '2023-01-02', 'cost': 800.0, 'quantity': 1, 'purchased_by': 'Mother', 'payment_method': 'Cash'}
[2024-07-01T12:28:17.836+0000] {logging_mixin.py:188} INFO - {'item_name': 'Headphones', 'date_of_purchase': '2023-01-03', 'cost': 150.0, 'quantity': 2, 'purchased_by': 'Father', 'payment_method': 'Card'}
[2024-07-01T12:28:17.837+0000] {logging_mixin.py:188} INFO - Error decoding JSON: Expecting value: line 2 column 1 (char 1)
[2024-07-01T12:28:17.840+0000] {conn.py:380} INFO - <BrokerConnection node_id=bootstrap-0 host=kafka:9092 <connecting> [IPv4 ('172.18.0.2', 9092)]>: connecting to kafka:9092 [('172.18.0.2', 9092) IPv4]
[2024-07-01T12:28:17.841+0000] {conn.py:1206} INFO - Probing node bootstrap-0 broker version
[2024-07-01T12:28:17.842+0000] {conn.py:410} INFO - <BrokerConnection node_id=bootstrap-0 host=kafka:9092 <connecting> [IPv4 ('172.18.0.2', 9092)]>: Connection complete.
[2024-07-01T12:28:17.947+0000] {conn.py:1268} INFO - Broker version identified as 2.6.0
[2024-07-01T12:28:17.947+0000] {conn.py:1269} INFO - Set configuration api_version=(2, 6, 0) to skip auto check_version requests on startup
[2024-07-01T12:28:17.951+0000] {taskinstance.py:441} INFO - ::group::Post task execution logs
[2024-07-01T12:28:17.952+0000] {taskinstance.py:2905} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 465, in _execute_task
    result = _execute_callable(context=context, **execute_callable_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 432, in _execute_callable
    return execute_callable(context=context, **execute_callable_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/baseoperator.py", line 400, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/operators/python.py", line 235, in execute
    return_value = self.execute_callable()
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.12/site-packages/airflow/operators/python.py", line 252, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/airflow/dags/script_kafka.py", line 80, in kafka_producer_script
    send_to_kafka(producer, kafka_topic, dataresult)
  File "/opt/airflow/dags/script_kafka.py", line 60, in send_to_kafka
    producer.send(topic, value=data)
  File "/home/airflow/.local/lib/python3.12/site-packages/kafka/producer/kafka.py", line 576, in send
    assert not (value is None and key is None), 'Need at least one: key or value'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Need at least one: key or value
[2024-07-01T12:28:17.968+0000] {taskinstance.py:1206} INFO - Marking task as FAILED. dag_id=purchase_dag, task_id=kafka_producer, run_id=manual__2024-07-01T12:28:14.869355+00:00, execution_date=20240701T122814, start_date=20240701T122817, end_date=20240701T122817
[2024-07-01T12:28:17.990+0000] {standard_task_runner.py:110} ERROR - Failed to execute job 488 for task kafka_producer (Need at least one: key or value; 593)
[2024-07-01T12:28:18.016+0000] {local_task_job_runner.py:240} INFO - Task exited with return code 1
[2024-07-01T12:28:18.053+0000] {taskinstance.py:3498} INFO - 0 downstream tasks scheduled from follow-on schedule check
[2024-07-01T12:28:18.055+0000] {local_task_job_runner.py:222} INFO - ::endgroup::
